<a href="https://colab.research.google.com/github/Varchala/NaturalLanguageProcessing_CSC8980/blob/main/hw3_nlp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**G Varchaleswari**

vganugapati1@student.gsu.edu

**002575320**


# 1. 
Using NLTK tokenize all documents, separated by polarity, remove stop words, and list
the top 20 most frequent tokens (and their counts) for the positive reviews, and the top
20 most frequent tokens (and their counts). What kind of things do you notice are
different between the two sets? (30 points)

In [171]:
# !tar -xvf /content/sample_data/review_polarity.tar -C /content/sample_data/

In [172]:
import pandas as pd
import numpy as np
import nltk
import os

In [173]:
def file_extract(data,dir):
  path = "/content/sample_data/txt_sentoken/"
  path += dir
  with os.scandir(path) as files:
    for file in files:
      if file.name.endswith(".txt"):
        with open(path+"/"+file.name) as review:
          data = data.append({'data':review.read(),'labels':dir,'name':file.name},ignore_index=True)
  return data

In [174]:
df = pd.DataFrame()
data_df=file_extract(df,"neg")
data_df = file_extract(data_df,"pos")

In [175]:
data_df

data labels             name
0     an attempt at florida film noir , palmetto fai...    neg  cv496_11185.txt
1     a hotshot lawyer gets an obviously guilty chil...    neg  cv265_11625.txt
2     chill factor is a carbon copy of speed with on...    neg  cv436_20564.txt
3     mugshot ( director/writer/cinematographer/edit...    neg  cv708_28539.txt
4     i never understood what the clich ? " hell on ...    neg  cv130_18521.txt
...                                                 ...    ...              ...
1995  one of the sweetest tales to ever be made , it...    pos  cv521_15828.txt
1996  after the terminally bleak reservoir dogs and ...    pos  cv822_20049.txt
1997  `run lola run' , a german import that gained a...    pos  cv710_22577.txt
1998  in october of 1962 the united states found its...    pos  cv771_28665.txt
1999  note : some may consider portions of the follo...    pos  cv162_10424.txt

[2000 rows x 3 columns]

In [176]:
data_df['labels'] = data_df['labels'].map({'pos': 1, 'neg': 0}).astype(int)
data_df

data  ...             name
0     an attempt at florida film noir , palmetto fai...  ...  cv496_11185.txt
1     a hotshot lawyer gets an obviously guilty chil...  ...  cv265_11625.txt
2     chill factor is a carbon copy of speed with on...  ...  cv436_20564.txt
3     mugshot ( director/writer/cinematographer/edit...  ...  cv708_28539.txt
4     i never understood what the clich ? " hell on ...  ...  cv130_18521.txt
...                                                 ...  ...              ...
1995  one of the sweetest tales to ever be made , it...  ...  cv521_15828.txt
1996  after the terminally bleak reservoir dogs and ...  ...  cv822_20049.txt
1997  `run lola run' , a german import that gained a...  ...  cv710_22577.txt
1998  in october of 1962 the united states found its...  ...  cv771_28665.txt
1999  note : some may consider portions of the follo...  ...  cv162_10424.txt

[2000 rows x 3 columns]

In [177]:
nltk.download('stopwords')
from nltk.corpus import stopwords
s_words = stopwords.words('english')
print(s_words)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'bo

# Code for removing stop words and counting the the top 20 most frequent words from positive and negative reviews.

In [178]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [179]:
def f_words(data,n):
  words = pd.DataFrame()
  for sent in data:
    for word in nltk.word_tokenize(sent):
      if word not in s_words:
        if not words.empty and word in words.word.values:
          words.loc[words.word == word, "count"] = words[words['word']==word]['count']+1
        else:
          words = words.append({'word':word,'count':1},ignore_index=True)
  return words.nlargest(n, 'count',keep='all').reset_index(drop=True)
print("Top 20 most frequent words from positive reviews \n")
print(f_words(data_df[data_df['labels']==1]['data'],20))
print("Top 20 most frequent words from negative reviews \n")
print(f_words(data_df[data_df['labels']==0]['data'],20))

Top 20 most frequent words from positive reviews 

      count       word
0   42448.0          ,
1   33714.0          .
2    9473.0         's
3    8494.0         ``
4    6039.0          )
5    6014.0          (
6    5186.0       film
7    2943.0        one
8    2775.0        n't
9    2497.0      movie
10   1713.0       like
11   1570.0          ?
12   1502.0          :
13   1231.0      story
14   1200.0       also
15   1190.0       good
16   1175.0       even
17   1171.0       time
18   1079.0      would
19   1067.0  character
Top 20 most frequent words from negative reviews 

      count   word
0   35269.0      ,
1   32162.0      .
2    9123.0     ``
3    8655.0     's
4    5742.0      )
5    5650.0      (
6    4257.0   film
7    3442.0    n't
8    3174.0  movie
9    2637.0    one
10   2201.0      ?
11   1832.0   like
12   1540.0      :
13   1381.0   even
14   1185.0  would
15   1126.0   good
16   1111.0   time
17   1056.0      !
18   1039.0    get
19   1019.0    bad


As few tokens like ',.? etc., are also considered in the document, the highest frequencies shown in the both sets are common for these. Also, even after removal of stopwords from NLTK, it still contains some stopwords that are relevant to the context in question i.e. "Movie reviews". It will contain some words like film, movie etc which are common in both. Word like "good" appeared 1190 times in positive reviews and 1126 times in negative reviews, this could be because good must have appeared as a combination of "not good" in negative reviews. Also, if we look at "n't" which appeared more in negative reviews than positive, most probable reason could be to express negative sentiments like "didn't like" or something like that.Bad appeared 1019 times in negative reviews and is not included in top 20 tokens of positive reviews.

# 2. 
Using the code from previous lectures, build 3 polarity classifiers using the following
parameters (20 points). Note: just train the models.
a) For training: use 50% of the positive dataset and 70% of the negative dataset. For
your model use: NaiveBayes with the TF-IDF vectorizer.
b) For training: use 50% of the negative dataset and 70% of the positive dataset. For
your model use: NaiveBayes with the TF-IDF vectorizer.
c) For training: use 25% of the negative dataset and 25% of the positive dataset. For
your model use: SVM with the TF-IDF vectorizer.


In [180]:
seed_value = 122

In [181]:
from sklearn.model_selection import train_test_split

In [182]:
def split_set(pos_per,neg_per):
  a_train,a_test = train_test_split(data_df[data_df['labels']==0], test_size=neg_per, random_state=seed_value)
  a_train1,a_test1 = train_test_split(data_df[data_df['labels']==1], test_size=pos_per, random_state=seed_value)
  a_train = a_train.append(a_train1)
  a_test = a_test.append(a_test1)
  return a_train,a_test

In [183]:
from sklearn.pipeline import make_pipeline
def train_model(aName,vecName, trainSet):
  model = make_pipeline(vecName, aName)
  model.fit(trainSet.data, trainSet.labels)
  return model

# 50% of the positive dataset and 70% of the negative dataset

In [184]:
train_a,test_a = split_set(0.50,0.30)

In [185]:
# train_a[train_a['labels']==0]

# NaiveBayes with the TF-IDF vectorizer

In [186]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import TfidfVectorizer

In [187]:
model_a = train_model(MultinomialNB(),TfidfVectorizer(),train_a)

# 50% of the negative dataset and 70% of the positive dataset

In [188]:
train_b,test_b = split_set(0.30,0.50)

# NaiveBayes with the TF-IDF vectorizer

In [189]:
model_b = train_model(MultinomialNB(),TfidfVectorizer(),train_b)

# 25% of the negative dataset and 25% of the positive dataset

In [190]:
train_c,test_c = split_set(0.75,0.75)

# SVM with the TF-IDF vectorizer

In [191]:
from sklearn.svm import SVC

In [192]:
model_c = train_model(SVC(),TfidfVectorizer(),train_c)

# 3.
Using the models from question 2, evaluate them on their individual rest of the dataset.
This is, for a) 50% positive and 30% negative, for b) 50% negative and 30% positive, and
for c) 75% negative and 75% positive. Calculate and show ONLY the following metrics for
each model: Accuracy, Precision, Recall, Macro F1-score. (15 points).

In [193]:
import sklearn.metrics as sm
def evaluate_model(test,model):
  pred = model.predict(test.data)
  acc = sm.accuracy_score(test.labels,pred)
  pr = sm.precision_score(test.labels,pred)
  re = sm.recall_score(test.labels,pred)
  f1 = sm.f1_score(pred, test.labels, average='macro')
  return acc,pr,re,f1

# a) 50% positive and 30% negative

In [194]:
accuracy,precision,recall,f1score=evaluate_model(test_a,model_a)

In [195]:
print("Model a \n")
print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1score)

Model a 

Accuracy: 0.3775
Precision: 1.0
Recall: 0.004
F1 Score: 0.2772081074608669


# b) 50% negative and 30% positive, 

In [196]:
accuracy,precision,recall,f1score=evaluate_model(test_b,model_b)

In [197]:
print("Model b \n")
print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1score)

Model b 

Accuracy: 0.375
Precision: 0.375
Recall: 1.0
F1 Score: 0.2727272727272727


# c) 75% negative and 75% positive

In [198]:
accuracy,precision,recall,f1score=evaluate_model(test_c,model_c)

In [199]:
print("Model c \n")
print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1score)

Model c 

Accuracy: 0.7553333333333333
Precision: 0.7338217338217338
Recall: 0.8013333333333333
F1 Score: 0.754814520859472


# 4) 
Using the model performance metrics from question 3, answer the following
questions. Please provide logical and intuitive rationale for your answers, simple
answers like: because it has the best score, will not be sufficient. (40 points):
a) What is the best performing model?
b) Why do you think this is the best performing model?
c) How does class imbalance play in determining polarity?
d) Do you think either more data or a better model is a better approach for this
kind of task?


**solution**

Undoubtedly, SVM is the best performing model with metric scores unparalleled to the ones of naive bayes.

But why is it so? If we consider how the algoritm works, Naive bayes works by calculating individual probabilities of features where one observation has no impact on the classification of another observation i.e. the fundamental assumption is that each feature makes an independent and equal contribution to the outcome. SVM is a geometrical non-probabilistic model which works relying on the interdependence of observations in the entire dataset. A hyperplane is drawn such that it separates the dataset into two classes and the orientation of the hyperplane is altered/ adjusted based on the support vector points or the observations that lie closer to the hyperplane. So what this tells us is that, polarity binary classification or sentiment analysis won't work best with probabilistic models as there is a feature to feature interdependence, till an extent, in doing predictions. If a text reads "The movie was not good and did not satisfy my interests", we can not consider ['good','satisfy'] and then predict it as positive. There is a relation between "good" and "not". So this problem is overcome by SVMs till an extent as opposed to Naive bayes algorithm.

Class imbalance introduces certain amount of bias in the data such that the model learns the majority class much better than the minority which results in improper polarity predictions for the minority classes. For example a model predicts the cancerous cells in a body where it was trained on a dataset with more negative classes than positive. When the model is tested on real-time data then predicting a potentially cancerous cell as negative will be a huge loss. This kind of scenarios should be kept in mind while preprocessing or having a dataset with the required number of classes.

Few believe that it is good to have bigger sample space for training better models. But how big is the "bigger"? This depends on the kind of model we are building and the context to which it is applied. To make it more sense, it is not always required to have humungous datasets such that we keep increasing the size. This will result in more processing time and a requirement of high-end resources to maintain the same. Techniques like cross validation give us very good estimates of how our model would perform on unseen data and then  based on which we can decide what should be the size of the data i.e., if a dataset is choosen such that applying cross validation results in poor accuracy scores on unseen data then there's a need to update the size of content of the dataset. 



# 5) 
Using NLTK and VADER, calculate the sentiment score for all documents in the
positive polarity. Calculate the polarity threshold needed (and reasonable) to have the
majority of the document labels match. Do the same for the negative class. Provide the
threshold needed, the reason why you think this threshold is reasonable, and the
accuracy percentage (how many documents are correctly labeled using this threshold).
(45 points):

In [200]:
nltk.download('vader_lexicon')
nltk.download('movie_reviews')
nltk.download('punkt')
!pip install twython

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!
[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [201]:
data_df_neg = data_df[data_df['labels']==0]

In [202]:
data_df_pos = data_df[data_df['labels']==1].reset_index(drop=True)

In [203]:
# data_df_pos

In [204]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer as SIA

sia = SIA()

# Negative polarity sentiment scores

In [205]:
polarity_df1 = pd.DataFrame()
for i in data_df_neg.data:
  polarity_df1 = polarity_df1.append(sia.polarity_scores(i),ignore_index=True)
data_df_neg = data_df_neg.join(polarity_df1,how="right")
data_df_neg

data  labels  ...    neu    pos
0    an attempt at florida film noir , palmetto fai...       0  ...  0.806  0.092
1    a hotshot lawyer gets an obviously guilty chil...       0  ...  0.730  0.116
2    chill factor is a carbon copy of speed with on...       0  ...  0.820  0.084
3    mugshot ( director/writer/cinematographer/edit...       0  ...  0.794  0.087
4    i never understood what the clich ? " hell on ...       0  ...  0.726  0.089
..                                                 ...     ...  ...    ...    ...
995  from dusk till dawn ( director/editor : robert...       0  ...  0.822  0.065
996  annie wilson ( cate blanchett ) , a widow who ...       0  ...  0.717  0.168
997  tommy lee jones chases an innocent victim arou...       0  ...  0.760  0.135
998  to sum the entire film " 54 " up in one senten...       0  ...  0.838  0.111
999  i'm going to start this review off with a hypo...       0  ...  0.806  0.076

[1000 rows x 7 columns]

In [206]:
# data_df_neg.iloc[1][0]

In [207]:
data_df_neg['compound'].describe()

count    1000.000000
mean        0.105364
std         0.902338
min        -0.999700
25%        -0.953200
50%         0.663700
75%         0.982100
max         0.999600
Name: compound, dtype: float64

Compound column obtained from polarity scores represents the aggregated sentiment score of the given text. This seemed the most relevant measure for calculating labels. Now, for the threshold, the statistics decribed in the above table shows that the values range from -0.99 (i.e. minimum) to 0.99 (i.e. maximum) with a mean of 0.105. But the median value is 0.663 which implies that the distribution is right skewed and if a value is choosen closer to the mean then we will end up with majority of unmatched document labels. As stated in the question, we need a threshold where *majority of the document labels match*.Choosing the 75th percentile or greater makes more sense as it will include more than 3/4th of the dataset. Also, choosing 0.99 seems like a rather more biased choice which will overestimate the chosen algorithm(here,in SentimentIntensityAnalyzer by Vader). 

**Choosen threshold is *0.983* and every value below the threshold will be labeled as *0***

In [208]:
data_df_neg[data_df_neg['compound']<0.983]

data  labels  ...    neu    pos
0    an attempt at florida film noir , palmetto fai...       0  ...  0.806  0.092
1    a hotshot lawyer gets an obviously guilty chil...       0  ...  0.730  0.116
2    chill factor is a carbon copy of speed with on...       0  ...  0.820  0.084
3    mugshot ( director/writer/cinematographer/edit...       0  ...  0.794  0.087
4    i never understood what the clich ? " hell on ...       0  ...  0.726  0.089
..                                                 ...     ...  ...    ...    ...
995  from dusk till dawn ( director/editor : robert...       0  ...  0.822  0.065
996  annie wilson ( cate blanchett ) , a widow who ...       0  ...  0.717  0.168
997  tommy lee jones chases an innocent victim arou...       0  ...  0.760  0.135
998  to sum the entire film " 54 " up in one senten...       0  ...  0.838  0.111
999  i'm going to start this review off with a hypo...       0  ...  0.806  0.076

[756 rows x 7 columns]

In [209]:
lb = []
for i in data_df_neg['compound']:
  if i<0.983:
    lb.append(0)
  else:
    lb.append(1)
data_df_neg['predicted_labels'] = lb
data_df_neg

data  ...  predicted_labels
0    an attempt at florida film noir , palmetto fai...  ...                 0
1    a hotshot lawyer gets an obviously guilty chil...  ...                 0
2    chill factor is a carbon copy of speed with on...  ...                 0
3    mugshot ( director/writer/cinematographer/edit...  ...                 0
4    i never understood what the clich ? " hell on ...  ...                 0
..                                                 ...  ...               ...
995  from dusk till dawn ( director/editor : robert...  ...                 0
996  annie wilson ( cate blanchett ) , a widow who ...  ...                 0
997  tommy lee jones chases an innocent victim arou...  ...                 0
998  to sum the entire film " 54 " up in one senten...  ...                 0
999  i'm going to start this review off with a hypo...  ...                 0

[1000 rows x 8 columns]

In [210]:
accuracy = (data_df_neg[data_df_neg['predicted_labels']==data_df_neg['labels']].shape[0]/data_df_neg.shape[0])*100
print("Accuracy percentage is:")
print(accuracy)

Accuracy percentage is:
75.6


# Positive polarity sentiment scores

In [211]:
polarity_df2 = pd.DataFrame()
for i in data_df_pos.data:
  polarity_df2 = polarity_df2.append(sia.polarity_scores(i),ignore_index=True)
data_df_pos = data_df_pos.join(polarity_df2,how="right")
data_df_pos

data  labels  ...    neu    pos
0    there seem to be two reactions to dark city . ...       1  ...  0.761  0.162
1    the relaxed dude rides a roller coaster \nthe ...       1  ...  0.827  0.109
2    weir is well-respected in hollywood for turnin...       1  ...  0.764  0.174
3    call me crazy , but i don't see saving private...       1  ...  0.749  0.132
4    is it just me , or have disney films gradually...       1  ...  0.734  0.207
..                                                 ...     ...  ...    ...    ...
995  one of the sweetest tales to ever be made , it...       1  ...  0.671  0.249
996  after the terminally bleak reservoir dogs and ...       1  ...  0.801  0.139
997  `run lola run' , a german import that gained a...       1  ...  0.817  0.138
998  in october of 1962 the united states found its...       1  ...  0.771  0.112
999  note : some may consider portions of the follo...       1  ...  0.802  0.095

[1000 rows x 7 columns]

In [212]:
data_df_pos['compound'].describe()

count    1000.000000
mean        0.647903
std         0.700739
min        -0.999600
25%         0.898600
50%         0.990650
75%         0.997200
max         0.999900
Name: compound, dtype: float64

Compound column describes the aggregated sentiment value of the given text. I want to base the classification, into labels, of the text on "compound" polarity score. The stats shown in the above column 3/4th of the dataset lies between 0.99 (maximum value) to 0.898600 (25th percentile). There are comparatively fewer observations/scores that fall under 0.8 ( less that 1/4th). In general, a compound score of 0.8 shows high positive polarity in a dataset although the choice of threshold purely depends on the kind of dataset used. This algorithm performs better for positive scores as opposed to the negative sentiment predictions. Potential reason could be that Vader doesn't penalize against negation that harshly when compared to other algorithms.

**Choosen threshold is *0.8* and every value above the threshold will be labeled as *1***

In [213]:
data_df_pos[data_df_pos['compound']>0.8]

data  labels  ...    neu    pos
0    there seem to be two reactions to dark city . ...       1  ...  0.761  0.162
1    the relaxed dude rides a roller coaster \nthe ...       1  ...  0.827  0.109
2    weir is well-respected in hollywood for turnin...       1  ...  0.764  0.174
3    call me crazy , but i don't see saving private...       1  ...  0.749  0.132
4    is it just me , or have disney films gradually...       1  ...  0.734  0.207
..                                                 ...     ...  ...    ...    ...
993  the question isn't why has grease been reissue...       1  ...  0.794  0.159
994  ingredients : neophyte lawyer , legal situatio...       1  ...  0.724  0.196
995  one of the sweetest tales to ever be made , it...       1  ...  0.671  0.249
996  after the terminally bleak reservoir dogs and ...       1  ...  0.801  0.139
997  `run lola run' , a german import that gained a...       1  ...  0.817  0.138

[792 rows x 7 columns]

In [214]:
lb = []
for i in data_df_pos['compound']:
  if i>0.8:
    lb.append(1)
  else:
    lb.append(0)
data_df_pos['predicted_labels'] = lb
data_df_pos

data  ...  predicted_labels
0    there seem to be two reactions to dark city . ...  ...                 1
1    the relaxed dude rides a roller coaster \nthe ...  ...                 1
2    weir is well-respected in hollywood for turnin...  ...                 1
3    call me crazy , but i don't see saving private...  ...                 1
4    is it just me , or have disney films gradually...  ...                 1
..                                                 ...  ...               ...
995  one of the sweetest tales to ever be made , it...  ...                 1
996  after the terminally bleak reservoir dogs and ...  ...                 1
997  `run lola run' , a german import that gained a...  ...                 1
998  in october of 1962 the united states found its...  ...                 0
999  note : some may consider portions of the follo...  ...                 0

[1000 rows x 8 columns]

In [215]:
accuracy = (data_df_pos[data_df_pos['predicted_labels']==data_df_pos['labels']].shape[0]/data_df_pos.shape[0])*100
print("Accuracy percentage is:")
print(accuracy)

Accuracy percentage is:
79.2


# Bonus 
(40 points): 
Repeat questions 2,3 and 4 removing all stopwords. Answer the
following questions: Did this change the results in any way? Why do you think so?

In [216]:
print(s_words)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [217]:
df_no_stopwords = pd.DataFrame()
w_str=""
s_count = 0
r_count = 0
for index,i in data_df.iterrows():
  for w in nltk.word_tokenize(i.data):
    if w not in s_words:
      w_str += w+" "
      r_count +=1
    else:
      s_count += 1
  df_no_stopwords = df_no_stopwords.append({'data':w_str,'labels':i['labels'],'name':i['name']},ignore_index=True)
  w_str=""
df_no_stopwords['labels'] = df_no_stopwords['labels'].astype('int32')
df_no_stopwords

data  ...             name
0     attempt florida film noir , palmetto fails fun...  ...  cv496_11185.txt
1     hotshot lawyer gets obviously guilty child mol...  ...  cv265_11625.txt
2     chill factor carbon copy speed one notable exc...  ...  cv436_20564.txt
3     mugshot ( director/writer/cinematographer/edit...  ...  cv708_28539.txt
4     never understood clich ? `` hell earth `` trul...  ...  cv130_18521.txt
...                                                 ...  ...              ...
1995  one sweetest tales ever made , 's wonderful li...  ...  cv521_15828.txt
1996  terminally bleak reservoir dogs brutally viole...  ...  cv822_20049.txt
1997  `run lola run ' , german import gained stellar...  ...  cv710_22577.txt
1998  october 1962 united states found brink nuclear...  ...  cv771_28665.txt
1999  note : may consider portions following text sp...  ...  cv162_10424.txt

[2000 rows x 3 columns]

In [218]:
# print(df_no_stopwords.iloc[0][0])
# print("\n")
# print(data_df.iloc[0][0])

In [219]:
def split_setStop(pos_per,neg_per):
  a_train,a_test = train_test_split(df_no_stopwords[df_no_stopwords['labels']==0], test_size=neg_per, random_state=seed_value)
  a_train1,a_test1 = train_test_split(df_no_stopwords[df_no_stopwords['labels']==1], test_size=pos_per, random_state=seed_value)
  a_train = a_train.append(a_train1)
  a_test = a_test.append(a_test1)
  return a_train,a_test

In [220]:
from sklearn.pipeline import make_pipeline
def train_stop_model(aName,vecName, trainSet):
  model = make_pipeline(vecName, aName)
  model.fit(trainSet.data, trainSet.labels)
  return model

# 50% of the positive dataset and 70% of the negative dataset

In [221]:
train_stop_a,test_stop_a = split_setStop(0.50,0.30)

# NaiveBayes with the TF-IDF vectorizer

In [222]:
model_stop_a = train_stop_model(MultinomialNB(),TfidfVectorizer(),train_stop_a)

# 50% of the negative dataset and 70% of the positive dataset

In [223]:
train_stop_b,test_stop_b = split_setStop(0.30,0.50)

# NaiveBayes with the TF-IDF vectorizer

In [224]:
model_stop_b = train_stop_model(MultinomialNB(),TfidfVectorizer(),train_stop_b)

# 25% of the negative dataset and 25% of the positive dataset

In [225]:
train_stop_c,test_stop_c = split_setStop(0.75,0.75)

# SVM with the TF-IDF vectorizer

In [226]:
model_stop_c = train_stop_model(SVC(),TfidfVectorizer(),train_stop_c)

# Using the **evaluate_model** method created in Q3

# a) 50% positive and 30% negative

In [227]:
accuracy,precision,recall,f1score=evaluate_model(test_stop_a,model_stop_a)

In [228]:
print("Model a \n")
print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1score)

Model a 

Accuracy: 0.3975
Precision: 1.0
Recall: 0.036
F1 Score: 0.3120133600725098


# b) 50% negative and 30% positive, 

In [229]:
accuracy,precision,recall,f1score=evaluate_model(test_stop_b,model_stop_b)

In [230]:
print("Model b \n")
print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1score)

Model b 

Accuracy: 0.3825
Precision: 0.3778337531486146
Recall: 1.0
F1 Score: 0.2860807422447991


# c) 75% negative and 75% positive

In [231]:
accuracy,precision,recall,f1score=evaluate_model(test_stop_c,model_stop_c)

In [232]:
print("Model c \n")
print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1score)

Model c 

Accuracy: 0.7553333333333333
Precision: 0.7343941248470012
Recall: 0.8
F1 Score: 0.7548442203133274


# Using Vader

In [233]:
sia_stop = SIA()

In [234]:
def polarity_calc(df_s):
  polarity_df1 = pd.DataFrame()
  for i in df_s.data:
    polarity_df1 = polarity_df1.append(sia_stop.polarity_scores(i),ignore_index=True)
  df_s = df_s.join(polarity_df1,how="right")
  return df_s


In [235]:
# df_no_stopwords

In [236]:
df_no_stop_neg = polarity_calc(df_no_stopwords[df_no_stopwords['labels']==0].reset_index(drop=True))
df_no_stop_pos = polarity_calc(df_no_stopwords[df_no_stopwords['labels']==1].reset_index(drop=True))

In [237]:
df_no_stop_neg.describe()

labels     compound          neg          neu          pos
count  1000.0  1000.000000  1000.000000  1000.000000  1000.000000
mean      0.0     0.178472     0.151750     0.673120     0.175106
std       0.0     0.884680     0.050057     0.056097     0.045794
min       0.0    -0.998900     0.030000     0.431000     0.000000
25%       0.0    -0.914775     0.118000     0.636000     0.142750
50%       0.0     0.796250     0.148000     0.675000     0.173000
75%       0.0     0.978725     0.183250     0.710000     0.202000
max       0.0     0.998900     0.569000     0.894000     0.345000

In [238]:
df_no_stop_pos.describe()

labels     compound          neg          neu          pos
count  1000.0  1000.000000  1000.000000  1000.000000  1000.000000
mean      1.0     0.635817     0.122094     0.670657     0.207236
std       0.0     0.700532     0.049264     0.056819     0.053313
min       1.0    -0.999200     0.010000     0.449000     0.066000
25%       1.0     0.899200     0.087000     0.634000     0.168000
50%       1.0     0.987100     0.116000     0.673000     0.204000
75%       1.0     0.996125     0.152000     0.707000     0.238000
max       1.0     0.999800     0.314000     0.838000     0.418000

In [239]:
def predict_labels_stop(df,theta,check):
  lb = []
  for i in df['compound']:
    if check == ">":
      if i>theta:
        lb.append(1)
      else:
        lb.append(0)
    elif check == "<":
      if i<theta:
        lb.append(0)
      else:
        lb.append(1)
  df['predicted_labels'] = lb
  return df

In [240]:
pred_spos_df = predict_labels_stop(df_no_stop_pos,0.85,">")
pred_spos_df

data  ...  predicted_labels
0    seem two reactions dark city . either love uni...  ...                 1
1    relaxed dude rides roller coaster big lebowski...  ...                 1
2    weir well-respected hollywood turning scripts ...  ...                 1
3    call crazy , n't see saving private ryan film ...  ...                 0
4    , disney films gradually lost appeal ? almost ...  ...                 1
..                                                 ...  ...               ...
995  one sweetest tales ever made , 's wonderful li...  ...                 1
996  terminally bleak reservoir dogs brutally viole...  ...                 1
997  `run lola run ' , german import gained stellar...  ...                 1
998  october 1962 united states found brink nuclear...  ...                 0
999  note : may consider portions following text sp...  ...                 0

[1000 rows x 8 columns]

In [241]:
accuracy = (pred_spos_df[pred_spos_df['predicted_labels']==pred_spos_df['labels']].shape[0]/pred_spos_df.shape[0])*100
print("Accuracy percentage is:")
print(accuracy)

Accuracy percentage is:
76.6


In [242]:
pred_sneg_df = predict_labels_stop(df_no_stop_neg,0.979,"<")
pred_sneg_df

data  ...  predicted_labels
0    attempt florida film noir , palmetto fails fun...  ...                 0
1    hotshot lawyer gets obviously guilty child mol...  ...                 0
2    chill factor carbon copy speed one notable exc...  ...                 0
3    mugshot ( director/writer/cinematographer/edit...  ...                 0
4    never understood clich ? `` hell earth `` trul...  ...                 0
..                                                 ...  ...               ...
995  dusk till dawn ( director/editor : robert rodr...  ...                 0
996  annie wilson ( cate blanchett ) , widow strugg...  ...                 0
997  tommy lee jones chases innocent victim around ...  ...                 0
998  sum entire film `` 54 `` one sentence , would ...  ...                 1
999  'm going start review hypothetical question . ...  ...                 0

[1000 rows x 8 columns]

In [243]:
accuracy = (pred_sneg_df[pred_sneg_df['predicted_labels']==pred_sneg_df['labels']].shape[0]/pred_sneg_df.shape[0])*100
print("Accuracy percentage is:")
print(accuracy)

Accuracy percentage is:
75.0


In [244]:
print("Total stop words:")
print(s_count)
print("Total words that are not stop words:")
print(r_count)
print("Percentage of stopwords:")
print(s_count*100/(s_count+r_count))
print("Percentage of non stopwords:")
print(r_count*100/(s_count+r_count))

Total stop words:
595282
Total words that are not stop words:
929208
Percentage of stopwords:
39.04794390255102
Percentage of non stopwords:
60.95205609744898


There is no significant difference between before stop words removal and after stop removal versions of models in Q2. There is a slight 0.01% increase of accuracy of models that are trained after stop words removal. The kind of dataset used probably plays the biggest role in determining the observed effects. One potential reason could be, stop words vary from document to document as what type of dataset it is. Dataset for movies reviews might contain words like film, movie,etc. which add no new information to the analysis. It is important to find such words with a clear analysis and then apply something like stopword removal than just applying a general pool of stop words. Another reason also could be about the size of dataset used and the number of rare words vs common words in the document. Overall understanding is that the Naive Bayes yielded slightly more better result after stopword removal than the SVM.

For Vader, it is observed that the accuracy has declined by over 2% for positive labels prediction for the choosen threshold which were based on the 25th percentile value of the compounded polarity score. Also, the chosen threshold for both is based on their individual statistics. For negative labels the accuracy has declined insignificantly by 0.6% which does not convey much about how helpful was the stop word removal. Rather, it seemed that stop word removal caused a negative impact on Vader. One potential reason could be that the removal of stopwords like but, then etc. resulted in misinterpretation of the context as Vader not just looks at the frequency of words vs its lexicons but also considers the context and the style of writing.